### ISSUE TITLE
Swagger UI Description allows HTML Injection (XSS by Design)

### ISSUE THREAD

**Issue by DinisCruz** [swagger-api/swagger-ui#830]
Hi, I was trying out the Swagger UI, and I noticed that the ```swagger.apiInfo.description``` value allows HTML, which could lead to an XSS injection (if that ```description``` string is maliciously manipulated)

Here is a unit test that replicates this issue (in Coffee-Script):

```coffee
QA_NWR_API = require './TM-QA-NWR-API'
app        = require '../../app'


describe.only 'swagger', ->
  page = QA_NWR_API.create(before, after)
  url = 'http://localhost:8002/'
  server = null

  before (done)->
    server = app.listen(8002);
    url.append('docs').GET (html)->
        html.assert_Is_String()
        done()

  after ->
    server.close()
    url.GET (html)->
        assert_Is_Null(html)
        done()

  it 'Issue 830 - Swagger UI Description allows HTML Injection (XSS by Design)', (done)->
    dynamicText = 'jQuery Injection - '.add_5_Random_Letters()
    javascript = "$(function() { $('#target').html('#{dynamicText}')})"
    app.swagger.apiInfo.description = "Description element allows: <h1 id='target'>...</h1> <script>#{javascript}</script>"
    page.chrome.open url + 'docs', ()->
      page.html (html,$)->
        $('#target').html().assert_Is(dynamicText)
        done()
```

... as executed in WebStorm: 

![image](https://cloud.githubusercontent.com/assets/656739/5754748/d4d33f60-9c94-11e4-85bd-96dd351089c3.png)

It looks like this is 'by design' since for example the [swagger-node-express](https://github.com/swagger-api/swagger-node-express/blob/master/sample-application/app.js) takes advantage of this 'feature'  to add a link to its description: https://github.com/swagger-api/swagger-node-express/blob/master/sample-application/app.js#L83-L90

![image](https://cloud.githubusercontent.com/assets/656739/5754675/e556eb08-9c93-11e4-8bf4-7b4d42d93c14.png)

I don't think that this is a high risk security issue, but it would be good to either:

a) add a note to the documentation (referencing this issue) with an alert to be careful with the ```swagger.apiInfo.description``` value, or
b) use markdown for the description value (which will allow features like links, without the html scripting capability)

**Comment by DinisCruz** [comment]
Here is another XSS Injection location

![image](https://cloud.githubusercontent.com/assets/656739/5764010/bc433188-9ce6-11e4-8cee-5a50902ae711.png)

**Comment by DinisCruz** [comment]
And a couple more on 

![image](https://cloud.githubusercontent.com/assets/656739/5765394/22abae3c-9cf1-11e4-957e-e1f9eb561efe.png)

which are rendered here:

![image](https://cloud.githubusercontent.com/assets/656739/5765405/2fbb0bae-9cf1-11e4-8239-0504a188b7f7.png)

**Comment by fehguy** [comment]
should be fixed in #862

**Comment by DinisCruz** [comment]
Hi, is there a test that confirms the XSS fix?

**Comment by mala** [comment]
@fehguy @mohsen1 
this is too bad way to fix xss https://github.com/swagger-api/swagger-ui/pull/862/files#diff-7b22c617cbad75cdd8123d16345c8e80R2699

example: ``<img src=x onerror=...>``
http://petstore.swagger.io/?url=http://api.ma.la/tmp/cors/swi/

I think that restriction of url param or attention about security is needed.

**Comment by mala** [comment]
ping @fehguy @mohsen1

**Comment by webron** [comment]
@mala - feel free to submit a PR.

**Comment by mala** [comment]
I can write a patch, but I don't know all function of swagger-ui, and I'm not using swagger-ui.
Please reopen this issue for now.

**Comment by IanLewis** [comment]
Please reopen this issue as it's a full on XSS vulnerability with the ability to exploit it just by having users click on URLs.

**Comment by webron** [comment]
While we do want to solve security issues with the UI, just reopening this issue won't help. The issue was opened on a specific XSS case. Opening an issue on 'you have XX vulnerabilities everywhere' doesn't help us isolate the problems and solving them, even if there are many of them, unless it comes with a clear method of checking and testing the whole project for such vulnerabilities.

If you'd like to help us further, please open specific issues pointing us to the various locations, or a general one with more information on how to map them out. Suggestions on how to fix the issues are also welcome, and PRs are great too.

**Comment by gmanfunky** [comment]
The specific case this issue described was not fixed with the initially referenced fix. https://github.com/swagger-api/swagger-ui/pull/862 added a 'sanitize' helper, but it was never put in use. 

@DinisCruz issue also goes on to describe that this isn't just a ```<script``` injection issue. 

To fix most XSS described in this issue, I suggest a combination of more liberal use of handlebar template escaping (here: https://github.com/swagger-api/swagger-ui/pull/1633)

API-Data can use Markdown to insert hyperlinks in descriptions if desired.

**Comment by nikhiljindal** [comment]
As pointed our earlier, the specific XSS case that this issue was filed for doesnt seem to be fixed yet.
To repro: http://petstore.swagger.io/?url=http://api.ma.la/tmp/cors/swi/ (you should not be getting that dialog)

I will recommend opening this issue until it is fixed.

**Comment by nikhiljindal** [comment]
One way to fix the reported XSS issue (http://petstore.swagger.io/?url=http://api.ma.la/tmp/cors/swi/) is to just remove the URL query param. Is that query param required?

(I can send a PR if removing it is fine)

**Comment by webron** [comment]
@nikhiljindal - it's required, yes. It's being used by quite a few users.

**Comment by DinisCruz** [comment]
btw, to exploit this you don't need user interaction since the payload can be delivered on any page the user visits (in an hidden image)

Ok, after looking at it for a little bit, this is actually a much worse vulnerability than I originally thought of, as it stands it looks like it will affect any website that exposes the swagger interface.

For example : http://kubernetes.io/kubernetes/third_party/swagger-ui/?url=http://api.ma.la/tmp/cors/swi/

![image](https://cloud.githubusercontent.com/assets/656739/15463958/6feb0a82-20c4-11e6-8588-d383b329304a.png)

That javascript code will now execute under the domain affected (in the case above http://kubernetes.io/ )

I'm sure there is a good way to do a Google Dork search for swagger exposed interfaces

**Comment by fehguy** [comment]
Fixed in 2.1.5

**Comment by DinisCruz** [comment]
great thanks

is there a test that confirms the fix?

can you point to the commit that fixed it?
